feat(ai): add avante.nvim for a better AI experience

georgeguimaraes opened the pull request:
I switched over to https://github.com/yetone/avante.nvim and I think this is the best experience in terms of UI and easiness to add patches to your code. 

I'm recommending it to everyone.

However, I couldn't make render-markdown to work on Avante. It seems that my spec here is running before the one defined in extras/lang/markdown.lua and getting overwritten even though I added `optional = true`. I'll try to dig in deeper, but if someone has any pointers, I'd appreciate it.

  - Fixes #4394

dpetka2001 commented:
Regarding `render-markdown`, `lazy.nvim` parses all the files alphabetically and adds to the plugin specs.

`lazyvim.plugins.extras.coding.avante` comes before `lazyvim.plugins.extras.lang.markdown`. 

A workaround would be to change the priority of the Markdown Extra. So, in this file https://github.com/LazyVim/LazyVim/blob/3dbace941ee935c89c73fd774267043d12f57fe2/lua/lazyvim/plugins/xtras.lua#L2-L11 you could add `["lazyvim.plugins.extras.lang.markdown"] = 10,` or whatever other value lower than 50 which is the default one. 

Not sure if it's the best solution. You might want to add it and also wait for Folke's review, who will obviously know better.

georgeguimaraes commented:
@dpetka2001 interesting, I thought that adding optional = true would also mean that it added the spec later in the chain. 
I've set the priority for markdown to 10, it worked. tks a lot!

dpetka2001 commented:
`optional` only means "add this spec if there's already a base spec present somewhere else". At the same file you will see somewhere at the end that the table `extras` is being sorted either by priority or if it's the same priority then alphabetically. Then it does `import = extra` for what is available in the table `extras`. So, `coding.avante` will be imported before `lang.markdown`.

dpetka2001 commented on lua/lazyvim/plugins/extras/coding/avante.lua:
```diff
@@ -0,0 +1,34 @@
+    end,
+    event = "VeryLazy",
+    -- we want to use head for now, since the releases are not frequent
+    version = false,
```
This is already default in [LazyVim](https://github.com/LazyVim/starter/blob/cb6349c8ae922d1c5745574f4d17b44f2731d451/lua/config/lazy.lua#L30), so it's not really needed?

dpetka2001 commented:
I saw that you removed the `lang.markdown` priority. Was that intentional? Did you find another way to make it work without that addition?

georgeguimaraes commented:
> I saw that you removed the `lang.markdown` priority. Was that intentional? Did you find another way to make it work without that addition?

@dpetka2001 bad rebase on my part

sei40kr commented:
AFAIS, the latest version of [avante.nvim](https://github.com/yetone/avante.nvim) only supports Neovim 0.10 or above, but currently LazyVim must support Neovim 0.9.0 or above.

> avante.nvim is currently only compatible with Neovim 0.10.1 or later. Please ensure that your Neovim version meets these requirements before proceeding.

dpetka2001 commented:
So, does `grug-far` and `rustaceanvim` and possibly other plugins I'm not aware of. Honestly, I really don't see any reason for users to still be using Neovim 0.9.* when 0.10.1 stable is out.

dpetka2001 commented:
@georgeguimaraes I just created #4506 for `render-markdown`. After a recent commit, it's not necessary any more to define both Lazy's `ft` and the plugin's `opts.file_types`. I opted in to get rid of `opts.file_types`, so that the plugin will still lazy-load. Maybe you would like to make corresponding changes to your PR and get rid of the `opts` function as a whole, since it's perfectly fine to only define the `Avante` filetype in Lazy's `ft`.

georgeguimaraes commented:
@dpetka2001 tks for the heads up!

dpetka2001 commented on lua/lazyvim/plugins/xtras.lua:
```diff
@@ -5,6 +5,7 @@ local prios = {
   ["lazyvim.plugins.extras.ui.edgy"] = 2,
   ["lazyvim.plugins.extras.lang.typescript"] = 5,
   ["lazyvim.plugins.extras.formatting.prettier"] = 10,
+  ["lazyvim.plugins.extras.lang.markdown"] = 10,
```
Not needed I believe? Since you removed the `opts` from `render-markdown` plugin spec? You should also test it out yourself in case I'm mistaken, but that's my assumption.

aarnphm commented:
> AFAIS, the latest version of [avante.nvim](https://github.com/yetone/avante.nvim) only supports Neovim 0.10 or above, but currently LazyVim must support Neovim 0.9.0 or above.
> 
> > avante.nvim is currently only compatible with Neovim 0.10.1 or later. Please ensure that your Neovim version meets these requirements before proceeding.

quick chime in, we can enable for 0.10 and above

aarnphm commented on lua/lazyvim/plugins/extras/coding/avante.lua:
```diff
@@ -0,0 +1,29 @@
+return {
+  {
+    "yetone/avante.nvim",
+    init = function()
+      require("avante_lib").load()
+    end,
```
You don't have to do this

georgeguimaraes replied:
hi! care to explain why? I wasn't sure by only reading the docs

dpetka2001 replied:
It seems to be needed when you want 
> For avante.tokenizers and templates to work, make sure to call require('avante_lib').load() somewhere when entering the editor. 

Not sure what `avante.tokenizers` and templates are, but it if a user wants to use them they suggest to do what you already do in this PR.

aarnphm replied:
It is already loaded in case of lazy.nvim

That doe is only valid for ppl who don't use lazy.nvim

aarnphm replied:
https://github.com/yetone/avante.nvim/blob/5c02a5d846a17b9f9db40cf64caa0bc75e7b13c1/lua/avante/init.lua#L236

dpetka2001 replied:
Oh, I see now. What I quoted was under the `Lua` section for when people do plain `require`s in their config without `lazy.nvim`. Ok that makes more sense. 

Thanks for clarifying

aarnphm replied:
I updated the docs accordingly sorry for any confusions.

georgeguimaraes commented:
@dpetka2001 Just an update: I had to resort to using both `ft` and `opts.file_types` for render-markdown; otherwise, the Avante window would give me a treesitter not loaded error when render-markdown tried to run there.

dpetka2001 commented:
Oh ok. I haven't used it, so i'll leave that to you. Then maybe you'd also like to leave the `lang.markdown` with higher priority to make sure it gets loaded first, sine you're also using render-markdown's `opts` here in this Extra?

aarnphm commented:
> @dpetka2001 Just an update: I had to resort to using both `ft` and `opts.file_types` for render-markdown; otherwise, the Avante window would give me a treesitter not loaded error when render-markdown tried to run there.

you can get the raw config from lazy and append the fite_type

leCoq42 commented:
Hi, I tried your config for avante, but it doesn't work well with blink-cmp(-copilot) unfortunately. And I think that folke is trying to replace nvim-cmp with blink.cmp (from what I make of the recent updates to lazyvim).

georgeguimaraes commented:
> Hi, I tried your config for avante, but it doesn't work well with blink-cmp(-copilot) unfortunately

@leCoq42 yeah, from what I understood, avante doesn't have support for blink yet.

@aarnphm is supporting blink on the roadmap for avante?

aarnphm commented:
yeah if I find time the next two weeks I can add blink support

exsesx commented:
Looking forward for this to be added to LazyExtras 🙏

georgeguimaraes commented:
LazyVim 14.0 shipped and removed dressing.nvim (as part of the move to fzf). Avante depends on it, so I added it on this PR:

 ```lua
    dependencies = {
      "stevearc/dressing.nvim",
    },
```

I also tested avante with blink instead of nvim-cmp (since blink is now the default). I have no problems whatsoever.

Mjb141 commented:
Any movement on this? This is currently my only gripe with LazyVim, Avante's experience is quite an improvement over current options.

dangra commented on lua/lazyvim/plugins/extras/coding/avante.lua:
```diff
@@ -0,0 +1,32 @@
+    ft = function(_, ft)
+      vim.list_extend(ft, { "Avante" })
+    end,
+    opts = function(_, opts)
+      opts.file_types = vim.list_extend(opts.file_types or {}, { "Avante" })
+    end,
```
No need for this since v7.8.0 https://github.com/MeanderingProgrammer/render-markdown.nvim/commit/cb9a5e2412d21c7a89627e0d6da5459acbc0eb9c

sei40kr replied:
@dangra 
I've tried to remove this line, but it didn't work since that commit doesn't support the case when `ft` is a function:
https://github.com/MeanderingProgrammer/render-markdown.nvim/commit/cb9a5e2412d21c7a89627e0d6da5459acbc0eb9c#diff-0a534dfa79afd0fc6225ed86d9dec59f8d76f56e559792a9ad3c9bd0f853ba04R22-R29

dangra reviewed (commented):
Are completions working for you with blink.cmp? Multiple people has reported issues in https://github.com/yetone/avante.nvim/issues/959

sei40kr commented:
@dangra 
For me, it's working. Here's my configuration:
<details>
<div>

```lua
  {
    "saghen/blink.cmp",
    dependencies = { "avante.nvim", "saghen/blink.compat" },
    opts = {
      sources = {
        compat = {
          "avante_commands",
          "avante_mentions",
        },
        default = {
          "avante_commands",
          "avante_mentions",
        },
        providers = {
          avante_commands = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 90,
            opts = {},
          },
          avante_files = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 100,
            opts = {},
          },
        },
      },
    },
    optional = true,
  },
```

</div>
</details>

As a side note, the `avante_files` mentioned in the README were not registered:
https://github.com/search?q=repo%3Ayetone%2Favante.nvim%20cmp_avante&type=code

sei40kr commented:
When opened in Snacks dashboard, the screen display is corrupted.
Other dashboard plugins may have the same problem.

FWIW, I just sent a PR to avante.nvim to change the filetype of the selected files buffer and it was merged: yetone/avante.nvim#1097

I then set up Edgy.nvim to get a consistent UI.
Here is my configuration:

<details>
<div>

```lua
  {
    "folke/edgy.nvim",
    opts = function(_, opts)
      opts.right = opts.right or {}

      -- FIXME: Somehow winbar is not set correctly
      table.insert(opts.right, {
        ft = "Avante",
        title = "Avante",
        size = { width = 0.3 },
      })
      -- FIXME: Somehow winbar is not set correctly
      table.insert(opts.right, {
        ft = "AvanteSelectedFiles",
        title = "Avante (Selected Files)",
        size = { height = 3 },
      })
      -- FIXME: Somehow Edgy keymaps are not set correctly
      table.insert(opts.right, {
        ft = "AvanteInput",
        title = "Avante (Input)",
        size = { height = 9 },
      })
    end,
    optional = true,
  },
```

</div>
</details>
As indicated in the comments, two of the windows do not have `winbar` set for some reason, and the other one does not have a keymap set.

NeckBeardPrince commented:
> @dangra
> 
> > Are completions working for you with blink.cmp?
> 
> For me, it's working. Here's my configuration:
> 
> ```lua
>   {
>     "saghen/blink.cmp",
>     dependencies = { "avante.nvim", "saghen/blink.compat" },
>     opts = {
>       sources = {
>         compat = {
>           "avante_commands",
>           "avante_mentions",
>         },
>         default = {
>           "avante_commands",
>           "avante_mentions",
>         },
>         providers = {
>           avante_commands = {
>             name = "avante_commands",
>             module = "blink.compat.source",
>             score_offset = 90,
>             opts = {},
>           },
>           avante_files = {
>             name = "avante_commands",
>             module = "blink.compat.source",
>             score_offset = 100,
>             opts = {},
>           },
>         },
>       },
>     },
>     optional = true,
>   },
> ```
> 
> As a side note, the `avante_files` mentioned in the README were not registered: https://github.com/search?q=repo%3Ayetone%2Favante.nvim%20cmp_avante&type=code

Sorry for what might be a dumb question, but that snippet in your blink.lua or avante.lua?

Mjb141 commented:
I'd really rather ask if LazyVim is ignoring/refusing this one than people start sharing configs that may or may not work for you / them / other versions / etc. 

@folke is this in consideration as an extra for Lazyvim? If not can this be closed to avoid the impression that this may be coming up, and we'll take the discussion to https://github.com/yetone/avante.nvim.

dpetka2001 commented:
There are currently 30 PRs open and maintainer seems like he has a lot of load and attention to the new `Snacks.picker` he just released trying to weed problems out. So, be a little patient? I'm sure he will share his thoughts when he has time.

Mjb141 commented:
> There are currently 30 PRs open and maintainer seems like he has a lot of load and attention to the new `Snacks.picker` he just released trying to weed problems out. So, be a little patient? I'm sure he will share his thoughts when he has time.

No part of my question implied any impatience or rush?!

This PR has been open for _four months_. It's not unreasonable to ask if this is not a priority or won't be implemented.

georgeguimaraes commented:
Not speaking for folke, of course, but I'll take a look on the updates from avante on the weekend and test it against LazyVim head. Rough weeks at work.

PatMulligan commented:
I am testing out the following avante.lua with good results, will keep testing and report back

This implements config for Blink.cmp from yetone/avante.nvim/pull/984

I am but a humble servant of the tech overlords 😆

```lua
return {
  {
    "yetone/avante.nvim",
    event = "VeryLazy",
    dependencies = {
      "stevearc/dressing.nvim",
    },
    opts = function(_, opts)
      -- Default configuration
      opts.hints = { enabled = false }

      -- File selector configuration
      --- @alias FileSelectorProvider "native" | "fzf" | "mini.pick" | "snacks" | "telescope" | string
      opts.file_selector = {
        provider = "fzf", -- Avoid native provider issues
        provider_opts = {},
      }

      -- Blink.cmp integration
      -- LSP score_offset is typically 60
      opts.providers = {
        avante_commands = {
          name = "avante_commands",
          module = "blink.compat.source",
          score_offset = 90, -- show at a higher priority than lsp
          opts = {},
        },
        avante_files = {
          name = "avante_files",
          module = "blink.compat.source",
          score_offset = 100, -- show at a higher priority than lsp
          opts = {},
        },
        avante_mentions = {
          name = "avante_mentions",
          module = "blink.compat.source",
          score_offset = 1000, -- show at a higher priority than lsp
          opts = {},
        },
      }

      opts.compat = {
        "avante_commands",
        "avante_mentions",
        "avante_files",
      }
    end,
    build = LazyVim.is_win() and "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" or "make",
  },
  {
    "MeanderingProgrammer/render-markdown.nvim",
    optional = true,
    ft = function(_, ft)
      vim.list_extend(ft, { "Avante" })
    end,
    opts = function(_, opts)
      opts.file_types = vim.list_extend(opts.file_types or {}, { "Avante" })
    end,
  },
  {
    -- support for image pasting
    "HakonHarnes/img-clip.nvim",
    event = "VeryLazy",
    opts = {
      -- recommended settings
      default = {
        embed_image_as_base64 = false,
        prompt_for_file_name = false,
        drag_and_drop = {
          insert_mode = true,
        },
        -- required for Windows users
        use_absolute_path = true,
      },
    },
  },
  {
    "folke/which-key.nvim",
    optional = true,
    opts = {
      spec = {
        { "<leader>a", group = "ai" },
      },
    },
  },
}

```

georgeguimaraes commented:
Tks @PatMulligan! I've been using the blink integration from Avante and it seems fine here as well.

exsesx commented:
Here’s the setup I ended up with:

`plugins/avante.lua`
```lua
return {
  {
    "yetone/avante.nvim",
    lazy = true,
    event = "VeryLazy",
    build = LazyVim.is_win()
        and "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false"
      or "make",

    opts = {
      provider = "copilot",
      auto_suggestions_provider = "copilot",
      copilot = {
        model = "claude-3.5-sonnet",
        temperature = 0,
        max_tokens = 8192,
      },
      hints = { enabled = false },
      file_selector = {
        provider = "snacks",
        provider_opts = {},
      },
    },

    dependencies = {
      {
        "MeanderingProgrammer/render-markdown.nvim",
        ft = function(_, ft)
          vim.list_extend(ft, { "Avante" })
        end,
      },
      {
        "folke/which-key.nvim",
        opts = {
          spec = {
            { "<leader>a", group = "ai" },
          },
        },
      },
    },
  },

  {
    "stevearc/dressing.nvim",
    lazy = true,
    opts = {
      input = { enabled = false },
      select = { enabled = false },
    },
  },

  {
    "saghen/blink.compat",
    lazy = true,
    opts = {},
    config = function()
      -- monkeypatch cmp.ConfirmBehavior for Avante
      require("cmp").ConfirmBehavior = {
        Insert = "insert",
        Replace = "replace",
      }
    end,
  },

  {
    "saghen/blink.cmp",
    lazy = true,
    opts = {
      sources = {
        default = { "avante_commands", "avante_mentions", "avante_files" },
        providers = {
          avante_commands = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 90, -- show at a higher priority than lsp
            opts = {},
          },
          avante_files = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 100, -- show at a higher priority than lsp
            opts = {},
          },
          avante_mentions = {
            name = "avante_mentions",
            module = "blink.compat.source",
            score_offset = 1000, -- show at a higher priority than lsp
            opts = {},
          },
        },
      },
    },
  },
}
```

You can set up your AI provider if needed (I use `copilot` by default), but overall, this setup works well for me.

P.S. The solution from @PatMulligan didn’t work for me. I believe the configuration for blink.cmp is misplaced, as it doesn’t support autocompletion for commands like `@file`, `@clear`, etc. (`avante_commands`).

PatMulligan commented:
Hey awesome thanks, that will help out with clarifying the README over @ yetone/avante.nvim/pull/984 

 I think we might be finally converging on a solution 🎉 
 
 @exsesx just one small typo I see in your `blink.cmp` config 

```diff
  {
    "saghen/blink.cmp",
    lazy = true,
    opts = {
      sources = {
        default = { "avante_commands", "avante_mentions", "avante_files" },
        providers = {
          avante_commands = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 90, -- show at a higher priority than lsp
            opts = {},
          },
          avante_files = {
-            name = "avante_commands",
+            name = "avante_files",
            module = "blink.compat.source",
            score_offset = 100, -- show at a higher priority than lsp
            opts = {},
          },
          avante_mentions = {
            name = "avante_mentions",
            module = "blink.compat.source",
            score_offset = 1000, -- show at a higher priority than lsp
            opts = {},
          },
        },
      },
    },
  },
```

PatMulligan commented on lua/lazyvim/plugins/extras/ai/avante.lua:
```diff
@@ -0,0 +1,68 @@
+    },
+    opts = function(_, opts)
+      -- Default configuration
+      opts.hints = { enabled = false }
+
+      -- File selector configuration
+      --- @alias FileSelectorProvider "native" | "fzf" | "mini.pick" | "snacks" | "telescope" | string
+      opts.file_selector = {
+        provider = "fzf",
+        provider_opts = {},
+      }
+
+      -- Blink.cmp integration
+      -- LSP score_offset is typically 60
+      opts.providers = {
+        avante_commands = {
+          name = "avante_commands",
+          module = "blink.compat.source",
+          score_offset = 90, -- show at a higher priority than lsp
+          opts = {},
+        },
+        avante_files = {
+          name = "avante_files",
+          module = "blink.compat.source",
+          score_offset = 100, -- show at a higher priority than lsp
+          opts = {},
+        },
+        avante_mentions = {
+          name = "avante_mentions",
+          module = "blink.compat.source",
+          score_offset = 1000, -- show at a higher priority than lsp
+          opts = {},
+        },
+      }
+      opts.compat = {
+        "avante_commands",
+        "avante_mentions",
+        "avante_files",
+      }
+    end,
```
```suggestion
    opts = {
      -- Default configuration
      hints = { enabled = false },

      ---@alias AvanteProvider "claude" | "openai" | "azure" | "gemini" | "cohere" | "copilot" | string
      provider = "claude", -- Recommend using Claude
      auto_suggestions_provider = "claude", -- Since auto-suggestions are a high-frequency operation and therefore expensive, it is recommended to specify an inexpensive provider or even a free provider: copilot
      claude = {
        endpoint = "https://api.anthropic.com",
        model = "claude-3-5-sonnet-20241022",
        temperature = 0,
        max_tokens = 4096,
      },

      -- File selector configuration
      --- @alias FileSelectorProvider "native" | "fzf" | "mini.pick" | "snacks" | "telescope" | string
      file_selector = {
        provider = "fzf", -- Avoid native provider issues
        provider_opts = {},
      },
    },
```

PatMulligan commented on lua/lazyvim/plugins/extras/ai/avante.lua:
```diff
@@ -0,0 +1,68 @@
+      }
+    end,
+    build = LazyVim.is_win() and "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" or "make",
+  },
```
```suggestion
  },
    {
    "saghen/blink.cmp",
    lazy = true,
    opts = {
      sources = {
        default = { "avante_commands", "avante_mentions", "avante_files" },
        compat = {
          "avante_commands",
          "avante_mentions",
          "avante_files",
        },
        -- LSP score_offset is typically 60
        providers = {
          avante_commands = {
            name = "avante_commands",
            module = "blink.compat.source",
            score_offset = 90,
            opts = {},
          },
          avante_files = {
            name = "avante_files",
            module = "blink.compat.source",
            score_offset = 100,
            opts = {},
          },
          avante_mentions = {
            name = "avante_mentions",
            module = "blink.compat.source",
            score_offset = 1000,
            opts = {},
          },
        },
      },
    },
  },

```

PatMulligan reviewed (commented):
added changes from @exsesx

also just passing opts as a table instead of a function since the latter seems unnecessary for this purpose. the extra claude spec might not be necessary

georgeguimaraes commented on lua/lazyvim/plugins/extras/ai/avante.lua:
```diff
@@ -0,0 +1,86 @@
+      },
+    },
+    build = LazyVim.is_win() and "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" or "make",
+  },
```
isn't this already the default?

PatMulligan replied:
Ah, I think you're correct. I copied it over from @exsesx and I misunderstood the purpose of the method name thinking it was "confirm" in the sense of, in case someone as changed it, make sure to confirm these are the values.

However, now looking, it just seems to be the value for when a completion is confirmed. In that case, I would imagine they are indeed the defaults! Not sure why they would be there, and can probably be removed then

georgeguimaraes commented:
I've added fzf-lua and blink.compat as dependencies. In a minimal setup, we need to explictly set them.

dpetka2001 commented:
I was wrong in my previous testing because in my personal configuration I still had a blink.cmp plugin spec. 

I did a fresh LazyVim installation and all works as it should. Only the corresponding `optional = true` to the non-avante plugin specs should be needed, so that `lazy.nvim` doesn't parse the spec and then lists it as disabled when a user has enabled nvim-cmp for example. The additional check for the picker would only be a cherry on the top if you'd like to configure automatically the file provider for avante, so just do what you like :smile:

Yorizel commented:
Whats is the corrente most up to date config for it?

kamushadenes commented:
Hello folks! Is there anything blocking this from getting merged?

iFarmGolems commented:
I'd also like to know 😇

saimon-moore commented:
Me too... :)

ronald2wing commented:
Me too... :)

Demianeen commented:
Hi guys, maybe just add thumbs up, since everyone who is subscribed to this PR gets notifications on each such message while it would also create a clutter for everyone who will read this PR

arbrandes commented:
I'm running what this PR suggests (plus the blink.compat monkey patch) and everything seems to work.  However, I'm having trouble figuring out how to use the auto-complete functionality (after toggling it on with `<leader>as`).  As I type, I get reasonable suggestions but only as what looks like [ghost-text](https://cmp.saghen.dev/configuration/completion#ghost-text), instead of as a snippet in the list of suggestions - which is what I would have expected.  Plus, the whole list of suggestions gets kinda clobbered.

Has anybody gotten this particular feature working properly?

mirsella commented:
what about adding the neotree shortcuts if the neotree extra is activated ?
would be cool to do it for snacks explorer too as its the default

80avin commented:
Sharing to help others:

I tried this but was continously facing errors while using `@codebase`.

What worked is to remove the [blink.cmp](https://github.com/Saghen/blink.cmp) & [blink.compat](https://github.com/Saghen/blink.compat) definition from this and using [blink-cmp-avante](https://github.com/Kaiser-Yang/blink-cmp-avante) as in https://github.com/AstroNvim/astrocommunity/blob/27584070d688d040a0d1de1c18eac0b3ade6b90d/lua/astrocommunity/completion/avante-nvim/init.lua#L3

dpetka2001 commented:
Pretty sure the dependencies as of the latest commit will get hard pulled in.

For example, `nvim-cmp` gets pulled in but is disabled because its plugin spec is explicitly disabled in the `blink.cmp` default Extra. But if it's guarded with `optional = true` then it won't even get pulled in. Same goes for `dressing.nvim`, `render-markdown.nvim` and `copilot.lua`. They all get hard pulled in as dependencies even though the explicit Extras are not enabled.

The only explicitly necessary dependencies are
```
"nvim-treesitter/nvim-treesitter",
    "nvim-lua/plenary.nvim",
    "MunifTanjim/nui.nvim",
```
and the rest are optional from Avante's README.

Also it might be worth looking into the [Astronvim example](https://github.com/AstroNvim/astrocommunity/blob/27584070d688d040a0d1de1c18eac0b3ade6b90d/lua/astrocommunity/completion/avante-nvim/init.lua) as already stated by @80avin. 

I don't use however the plugin and can't be of further help regarding its configuration.

surmish commented:
@georgeguimaraes could you add `optional = true` for the plugins in the dependency list?

80avin commented:
FYI, here is my avante config

<details>

<summary>Click to see my avante.nvim config</summary>


```lua
return {
  {
    "yetone/avante.nvim",
    event = "VeryLazy",
    dependencies = {
      {
        "ravitemer/mcphub.nvim",
        cmd = "MCPHub",
        build = "pnpm install -g mcp-hub@latest",
        opts = {},
        keys = {
          { "<leader>am", "<cmd>MCPHub<cr>", mode = { "n" }, desc = "MCP Hub" },
        },
      },
    },
    opts = {
      hints = { enabled = false },
      provider = "copilot", -- Recommend using Claude
      providers = {
        copilot = {
          model = "gpt-4o",
          endpoint = "https://api.githubcopilot.com",
          allow_insecure = false,
          timeout = 10 * 60 * 1000,
          extra_request_body = { temperature = 0 },
          max_completion_tokens = 1000000,
          reasoning_effort = "high",
        },
      },
      auto_suggestions_provider = nil,
      behaviour = {
        auto_suggestions = false,
      },

      -- File selector configuration
      file_selector = {
        provider = "snacks", -- Avoid native provider issues
        provider_opts = {},
      },
      selector = { provider = "snacks" },
      system_prompt = function()
        local hub = require("mcphub").get_hub_instance()
        return hub and hub:get_active_servers_prompt() or ""
      end,
      custom_tools = function()
        return {
          require("mcphub.extensions.avante").mcp_tool(),
        }
      end,
      extensions = {
        avante = {
          make_slash_commands = true,
        },
      },
      input = { provider = "snacks" },
    },
    build = LazyVim.is_win() and "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" or "make",
  },
  {
    "Kaiser-Yang/blink-cmp-avante",
    lazy = true,
    specs = {
      {
        "saghen/blink.cmp",
        optional = true,
        opts = {
          sources = {
            default = { "avante" },
            providers = { avante = { module = "blink-cmp-avante", name = "Avante" } },
          },
        },
      },
    },
  },
  {
    "MeanderingProgrammer/render-markdown.nvim",
    optional = true,
    ft = function(_, ft)
      vim.list_extend(ft, { "Avante" })
    end,
    opts = function(_, opts)
      opts.file_types = vim.list_extend(opts.file_types or {}, { "Avante" })
    end,
  },
  {
    "folke/which-key.nvim",
    optional = true,
    opts = {
      spec = {
        { "<leader>a", desc = "ai" },
      },
    },
  },
}
```

</details>

PatMulligan commented:
Thanks @80avin ! This works great. The only thing is my avante wasn't building and I had to run `cd ~/.local/share/nvim/lazy/avante.nvim && make` manually.

PatMulligan commented:
> > Thanks @80avin ! This works great. The only thing is my avante wasn't building and I had to run `cd ~/.local/share/nvim/lazy/avante.nvim && make` manually.
> 
> That's weird. Do you want to describe your situation better, so that others facing this can benefit ?

Here are some more details:

OS: Pop!_OS 22.04 LTS x86_64`

NVIM v0.11.2
Build type: Release
LuaJIT 2.1.1741730670

Here is the specific error.
```
Error detected while processing TextChangedI Autocommands for "<buffer=7>":
Error executing lua callback: ...ug/.local/share/nvim/lazy/avante.nvim/lua/avante/llm.lua:215: Make sure to build avante (missing avante_templates)
stack traceback:
	[C]: in function 'error'
	...g/.local/share/nvim/lazy/avante.nvim/lua/avante/path.lua:215: in function 'get_templates_dir'
	...ug/.local/share/nvim/lazy/avante.nvim/lua/avante/llm.lua:215: in function 'generate_prompts'
	...ug/.local/share/nvim/lazy/avante.nvim/lua/avante/llm.lua:401: in function 'calculate_tokens'
	...local/share/nvim/lazy/avante.nvim/lua/avante/sidebar.lua:2140: in function 'cb'
	...local/share/nvim/lazy/avante.nvim/lua/avante/sidebar.lua:2584: in function 'get_generate_prompts_options'
	...local/share/nvim/lazy/avante.nvim/lua/avante/sidebar.lua:2139: in function 'show_input_hint'
	...local/share/nvim/lazy/avante.nvim/lua/avante/sidebar.lua:2922: in function <...local/share/nvim/lazy/avante.nvim/lua/avante/sidebar.lua:2921>
```
I used the same config as you mentioned, just swapped out copilot for claude. Also tried moving the build command to the top: 

```
return {
  {
    "yetone/avante.nvim",
    build = function()
      -- conditionally use the correct build system for the current OS
      if vim.fn.has("win32") == 1 then
        return "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false"
      else
        return "make"
      end
    end,
    event = "VeryLazy",
    dependencies = {
      {
        "ravitemer/mcphub.nvim",
        cmd = "MCPHub",
        build = "pnpm install -g mcp-hub@latest",
        opts = {},
        keys = {
          { "<leader>am", "<cmd>MCPHub<cr>", mode = { "n" }, desc = "MCP Hub" },
        },
      },
    },
    opts = {
      hints = { enabled = false },
      provider = "openai",
      -- for example
      providers = {
        copilot = {
          model = "gpt-4o",
          endpoint = "https://api.githubcopilot.com",
          allow_insecure = false,
          timeout = 10 * 60 * 1000,
          extra_request_body = { temperature = 0 },
          max_completion_tokens = 1000000,
          reasoning_effort = "high",
        },
        claude = {
          endpoint = "https://api.anthropic.com",
          model = "claude-sonnet-4-20250514",
          timeout = 30000, -- Timeout in milliseconds
          extra_request_body = {
            temperature = 0.75,
            max_tokens = 20480,
          },
        },
      },
      auto_suggestions_provider = nil,
      behaviour = {
        auto_suggestions = false,
      },

      -- File selector configuration
      file_selector = {
        provider = "snacks", -- Avoid native provider issues
        provider_opts = {},
      },
      selector = { provider = "snacks" },
      system_prompt = function()
        local hub = require("mcphub").get_hub_instance()
        return hub and hub:get_active_servers_prompt() or ""
      end,
      custom_tools = function()
        return {
          require("mcphub.extensions.avante").mcp_tool(),
        }
      end,
      extensions = {
        avante = {
          make_slash_commands = true,
        },
      },
      input = { provider = "snacks" },
    },
  },
  {
    "Kaiser-Yang/blink-cmp-avante",
    lazy = true,
    specs = {
      {
        "saghen/blink.cmp",
        optional = true,
        opts = {
          sources = {
            default = { "avante" },
            providers = { avante = { module = "blink-cmp-avante", name = "Avante" } },
          },
        },
      },
    },
  },
  {
    "MeanderingProgrammer/render-markdown.nvim",
    optional = true,
    ft = function(_, ft)
      vim.list_extend(ft, { "Avante" })
    end,
    opts = function(_, opts)
      opts.file_types = vim.list_extend(opts.file_types or {}, { "Avante" })
    end,
  },
  {
    "folke/which-key.nvim",
    optional = true,
    opts = {
      spec = {
        { "<leader>a", desc = "ai" },
      },
    },
  },
}
```

The solution on linux is to go to the avante.nvim local folder and run `cd ~/.local/share/nvim/lazy/avante.nvim && make`, or i suppose if it's on windows and run into a similar problem you would run `powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false`

80avin commented:
@PatMulligan my `build` option isn't a function but an expression evaluated when this file runs.

If it is a function, then the function should take care of building by executing `make` or `powershell` instead of returning it.
ref: https://lazy.folke.io/developers#building
